services/engine.go

```diff
@@ -0,0 +1,142 @@
+package services
+
+import (
+	"errors"
+)
+
+type Marketer interface {
+	PlaceOrder(o Order) error
+	CancelOrder(oID uint) error
+}
+
+/*
+TradeExchangeEngine is implementation of Marketer
+*/
+
+var _ Marketer = NewExchangeEngine(nil, nil, nil, nil, nil)
+
+func NewExchangeEngine(
+	u UserExistentChecker,
+	o Trader,
+	m Matcher,
+	ex Exchanger,
+	l []Listener,
+) *TradeExchangeEngine {
+	return &TradeExchangeEngine{
+		u, o, m, ex, l,
+	}
+}
+
+type TradeExchangeEngine struct {
+	userExistentChecker UserExistentChecker
+	trader              Trader
+	matcher             Matcher
+	exchanger           Exchanger
+	listeners           []Listener
+}
+
+func (e TradeExchangeEngine) PlaceOrder(o Order) error {
+	if e.userExistentChecker.Check(o.UserID) {
+		return errors.New("user not found")
+	}
+
+	e.trader.Place(o)
+	orders := e.trader.GetOrders()
+	e.publishEvent(Event{Name: "order has placed", Payload: o})
+
+	matches := e.matcher.Match(&o, orders)
+	e.exchanger.Exchange(o.UserID, matches, func(c BalanceChange) {
+		e.publishEvent(Event{Name: "balance has changed", Payload: c})
+	})
+
+	return nil
+}
+
+func (e TradeExchangeEngine) CancelOrder(oID uint) error {
+	o, canceled := e.trader.Cancel(oID)
+	if !canceled {
+		return errors.New("order can not be cancel ")
+	}
+
+	e.publishEvent(Event{Name: "order has canceled", Payload: o})
+	return nil
+}
+
+func (e TradeExchangeEngine) publishEvent(event Event) {
+	for _, l := range e.listeners {
+		l.Listen(event)
```

Comment:
The event handler and the engine are running on the same thread. It's better to leave the engine run alone.